common/context/src/main/java/io/helidon/common/context/ListContext.java

```diff
@@ -0,0 +1,221 @@
+                return parent.get(classifier, type);
+            } else {
+                return Optional.ofNullable(result);
+            }
+        } else {
+            if (parent != null) {
+                return parent.get(classifier, type);
+            } else {
+                return Optional.empty();
+            }
+        }
+    }
+
+    private interface RegisteredItem<T> {
+        T get();
+
+        Class<T> getType();
+    }
+
+    private static class ClassifiedRegistry {
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+        private final List<RegisteredItem<?>> content = new ArrayList<>();
+
+        // we actually want to do an instance equality
+        @SuppressWarnings("ObjectEquality")
+        private void registerItem(RegisteredItem<?> item) {
+            Lock l = lock.writeLock();
+            try {
+                Class<?> c = item.getType();
+                l.lock();
+                for (int i = 0; i < content.size(); i++) {
+                    RegisteredItem<?> reg = content.get(i);
+                    if (c == reg.getType()) {
+                        content.remove(i);
+                        break;
+                    }
+                }
+                content.add(item);
+            } finally {
+                l.unlock();
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        <T> void register(T instance) {
+            Objects.requireNonNull(instance, "Parameter 'instance' is null!");
+            registerItem(new RegisteredInstance(instance));
+        }
+
+        @SuppressWarnings("unchecked")
+        <T> void supply(Class<T> type, Supplier<T> supplier) {
+            Objects.requireNonNull(type, "Parameter 'type' is null!");
+            Objects.requireNonNull(supplier, "Parameter 'supplier' is null!");
+            registerItem(new RegisteredSupplier(type, supplier));
+        }
+
+        <T> T get(Class<T> type) {
+            Objects.requireNonNull(type, "Parameter 'type' is null!");
+            Lock l = lock.readLock();
+            try {
+                l.lock();
+                for (int i = content.size() - 1; i >= 0; i--) {
+                    RegisteredItem<?> item = content.get(i);
+                    if (type.isAssignableFrom(item.getType())) {
+                        return type.cast(item.get());
+                    }
+                }
+            } finally {
+                l.unlock();
+            }
+            return null;
+        }
+    }
+
+    private static class RegisteredSupplier<T> implements RegisteredItem<T> {
+        private final Class<T> type;
+        private final Supplier<T> supplier;
+        private volatile boolean missing = true;
+        private volatile T instance;
+
+        RegisteredSupplier(Class<T> type, Supplier<T> supplier) {
+            this.type = type;
+            this.supplier = supplier;
+        }
+
+        @Override
+        public T get() {
+            if (missing) {
```

Comment:
Is it your intention here that it will continue to retry on subsequent calls if missing?

Reply:
This implementation was moved from common http module - it comes from webserver. I did not modify the implementation at all (e.g. this is how our context works now in webserver)

Reply:
I probably do not see the issue. If missing, missing is set to `false` - so next time the instance is returned directly.
This is a lazy supplier that caches the supplied value and as I see it, it is written correctly